### DISCUSSION TITLE
Building Ultra Low Power with MicroPython

### DISCUSSION THREAD

**Discussion by volkerjaenisch** [micropython/micropython#12111] [General]
How to achieve the lowest possible power consumption with MicroPython?

Probably there is such a topic already out there - then please give me a pointer.

The above question seems a bit naive since it depends on ... a lot of things.

Motivation:
The project that made me open this topic is building a solar powered low cost digital moisture sensor with WiFi (or LoRa) out of widely available components.

The components are IMHO straight forward:
* ESP32 Dev-Board
* Digital-Moisture sensor [](https://github.com/BeFlE/SoMoSe)
* LiPo Accumulator
* Solar panel
* LiPo solar loader

Problems:
* Hardware: 
  * There are ESP boards out there with voltage regulators that draw 10-100 times more current that the ESP in deep sleep
  * Sensors are drawing a lot of power if kept unchecked
* Software:
  * All ESP Controllers are featuring a machine.deepsleep functionality, but each bit of the ESP32 hardware has to be shut down manually by the programmer before entering deepsleep. This fact is not documented.
  * And it is not really clear how to invoke a hardware component and how to set it to sleep. For most hardware components micropython features a "init" method, which is complemented by a "deinit" method. But there is a lack of documentation if e.g. the "deinit" method really shuts the hardware down, or not.
    * And there is e.g. no method to "deinit"  for the I2C bus.

Status:
With my current code I am able to quench the current down to 0.4mA on a DFRobot ESP8266 Firebeetle in deep-sleep. But is this the limit? 

Goals:
Let us document what hardware in combination with what code achieves the minimal current.

Cheers,
Volker

**Comment by davefes** [comment]
I have been moving all my solar-powered remote sensors to LoRa.
Also, I replace the AMS1117 with HZ73333 (4-8uA Quiescent Current)

> All ESP Controllers are featuring a machine.deepsleep functionality, but each bit of the ESP32 hardware has to be shut down manually by the programmer before entering deepsleep. This fact is not documented.

This is news to me ... which bits of hardware stay powered-up?

Have you read this:
https://github.com/glenn20/upy-esp32-experiments/blob/main/ESPNowvsWifiEnergyUsage/README.md

**Comment by karfas** [comment]
> each bit of the ESP32 hardware has to be shut down manually by the programmer before entering deepsleep. This fact is not documented.
>And it is not really clear how to invoke a hardware component and how to set it to sleep.


This is not true. The micropython docs are terse, yes, but other sources of information are
-  the ESP-IDF documentation
-  data sheets
-  source code.

I really have no idea why so many people seem to expect that uPy docs should cover every possible hardware configuration under the sun.
Programming bare metal (or freeRTOS) **does** require at least minimal knowledge about hardware, protocols, etc.

**Comment by peterhinch** [comment]
@volkerjaenisch I suggest looking at [the FeatherS3](https://esp32s3.com/feathers3.html). The ESP32-S3 has extremely low deepsleep current (unlike ESP32) and the board is well designed. I haven't managed to get a measurement of deepsleep current for the entire board powered from LiPo, but it's below 1μA. See also [this doc](https://github.com/glenn20/upy-esp32-experiments/tree/main/ESPNowvsWifiEnergyUsage) from @glenn20 who has done a lot of work on minimising energy use on boot.

**Reply by volkerjaenisch** [reply]
Thanks a lot for this valuable and constructive information. The Adafruit hardware was hard to get in Germany, or I did not found a good dealer, so I had stayed with my ESP32s. Will look out for a FeatherS3 - it is truly a jewel.
Cheers,
Volker

**Reply by glenn20** [reply]
Yes, Sion, the unexpected maker has put a lot of work into power management on his boards and the FeatherS3 or TinyS3 are great examples of that (along with the low-power advantages of the ESP32-S3 device).

Re. powering off the hardware, be careful of how you interpret the docs. The ESP-IDF documentation (for esp32 devices - not esp8266) clearly states that the wifi radio should be shutdown before sleep, but that is primarily to ensure it will function correctly after you wake - I **believe** that the radio will be powered off in any case (there may be some pathological behaviour where this is not the case from user reports - but I haven't seen it). On micropython the wifi radio is powered off (`esp_wifi_stop()`) if **both** the STA_IF and AP_IF are set `active(False)`. I can't speak from personal experience on the other peripherals, but I believe the default behaviour is that they are also shutdown - unless programmed otherwise.

There is a lot of documentation on the behaviour of the esp32 peripherals during power off in the esp-idf documentation. Micropython can't and shouldn't try to replicate all that documentation. Also micropython doesn't provide access to all the capabilities available through the IDF, but Issues can be raised if users need access to some of those capabiltiies (or, even better, PRs).

I know that you are using an esp8266 device, but I think the best opportunities for achieving very low power operations are probably with the ESP32 devices (partly due to the availability of modern well-optimised boards like the feathers3 mentioned by Peter) and much more comprehensiive documentation (though often ambiguous).

**Reply by glenn20** [reply]
I just ran a lightsleep and deepsleep test on my FeatherS3 with the PPK2:
- Normal idle operation (wifi off): **46mA**
- Light sleep:  **1.3mA**
- Deep sleep: **0.043mA**

I configured the PPK2 as a power source providing 3.8V through the lipo battery connector.

Tested with the following code:
```py
import machine
import network
import time
import secrets

# Wifi on and connected for 3 seconds
sta = network.WLAN(network.STA_IF)
sta.active(True)
sta.connect(secrets.ssid, secrets.key)
time.sleep(3)
sta.active(False)

# Wifi off for 1 second
time.sleep(1)

# Lightsleep for 2 second
machine.lightsleep(2000)

# Deepsleep for 2 seconds
machine.deepsleep(2000)
```

**Reply by volkerjaenisch** [reply]
Yesterday I measured (utilizing an INA219) the current of the DFrobot ESP8266 while running my SoMoSe demo [mqtt-app](https://github.com/Inqbus/somose/tree/main/mqtt_app/ESP8266) . The Board was powered with a 300mAH LiPo during the measurement.

The mqtt demo:
* powers up the SoMoSe sensor, does tree measurements (1 sec apart) and powers it down.
* Powers up WLAN, connects to the WLAN router.
* Connects to the MQTT-Server, sends the data, disconnects.
* Powers down WLAN
* deep sleep for 120 seconds.

![amperes](https://github.com/micropython/micropython/assets/2539336/90286fac-81bb-4ef4-93a6-989c8f883327)

The resolution of the INA219 is 12 bit, only, therefore the coarse Y-resolution in the image.

IMHO the DFrobot board is not that bad. Around 0.4 mA during deep sleep. 
OK this is an OOM of the FeatherS3, but I am still happy to get that low.

Cheers, 
Volker

**Reply by glenn20** [reply]
Good results. The firebeetle is known for having good deepsleep current - much better than most of it's contemporaries. When you calaculate the battery life at 0.4mA - that's still impressive.

**Reply by volkerjaenisch** [reply]
Here the whole picture:

Firebeetle 8266:

![amperes](https://github.com/micropython/micropython/assets/2539336/66c4fa36-1836-42f9-814d-337211d1ec2a)

Firebeetle ESP32 V4.0.

![amperes](https://github.com/micropython/micropython/assets/2539336/1d7ae1b6-beff-4a21-9aa5-94f49aa7f415)

Easy to identify: Booting up with 40-50 mA. Then baseline [time.sleep(10)] then the operation as before.

Here I did a 10 second long measurements to distinguish them better from the WLAN/MQTT-Part. 

The deep sleep consumption is 0.3mA in the mean, which is a bit lower than the ESP8266.

Interesting is here that the Sensor draws a lower current 50-60mA than on the ESP8266 Board (80mA). The sensor is supplied via a GPIO-Pin (for powering off in deep-sleep, and to reset it if hanging). Is it possible that the ESP32 GPIOs are weaker?

The WLAN utilizes a bit more power but my feeling is that the connection time is much faster - which may be a good tradeoff.

Any interpretations welcome.

**Reply by volkerjaenisch** [reply]
Ahh! Silly me. I should make a base line measurement first. Stay tuned.

**Reply by davefes** [reply]
30uA?  deepsleep.  ESP32 drive level 

>         Pin.DRIVE_0: 5mA / 130 ohm
> 
>         Pin.DRIVE_1: 10mA / 60 ohm
> 
>         Pin.DRIVE_2: 20mA / 30 ohm (default strength if not configured)
> 
>         Pin.DRIVE_3: 40mA / 15 ohm

What minimum voltage does the sensor need to see and do you get that on the ESP32 with the default drive strength?

**Reply by volkerjaenisch** [reply]
@davefes 
Here the sensor datasheet.

[SoMoSe moisture sensor](https://github.com/BeFlE/SoMoSe)

But it was my fault, the base line was missing. The 80mA were the complete consumption and not that of the sensor.
Please stay tuned, the base line measurements are coming in any minute.

**Reply by davefes** [reply]
3V3 at 20mA, which is the default drive setting.  E drop = 0.02A * 30 Ohms = 0.6V drop.  I'd check the voltage actually on the sensor.

**Reply by volkerjaenisch** [reply]
Dear @davefes 

Please be gentle with me. My electronics knowledge is sparse and English not my mother language, so many terms you are familiar with are a riddle for me.

1) What do you mean by "drop"?

2) The sensor needs a min. of 3.0 V for the resonant circuit (according to the producer).

3) You are completely right that I should measure the sensor characteristics at the sensor :-).  But my aim was to first understand the overall power consumption.

4) I have to clear my other workbench to get a working environment. Current measurements are a bit improvised :-)

![workspace](https://github.com/micropython/micropython/assets/2539336/2693fc14-f7ba-4826-942b-b5236c5054b7)

5) Let us discuss the sensor and its powering via GPIO in another thread. I will open it when I get a more robust working environment. I may also invite the developer of the sensor so we can use his first hand knowledge. He has lots of customers struggling with ESP32 and low power. 

6) @all: Many thanks for the support 

7) Disclaimer: This is a spare time project. I am the maintainer of the SoMoSe-Micropython open source driver since I own one and needed a driver.

Cheers,
Volker

**Reply by davefes** [reply]
1) at the default drive current setting they say the source impedance (resistance in this case) is 30 Ohms.  So, when you try to draw 20ma you end up with a decreased GPIO drive voltage.  0.6V "lost" across this internal source resistance takes you below the 3.0V min voltage for the sensor.  Even at drive level 3 you are going to "lose" or drop 0.3V, which is at least still in spec.  Measure the actual voltage on the sensor and verify it is about 2.7V, indicating a 20mA current consumption.  Then try drive level 3 and then I can suggest another approach in your new thread.

**Reply by volkerjaenisch** [reply]
Dear @davefes
Thank you so much for your explanation. Now I got it.

I will open up a new thread concerning the power supply of the SoMoSe Sensor.  

**To the gods of this forum: Is it OK to discuss such a specific electronic topic here?**

Cheers,
Volker

**Reply by jimmo** [reply]
> Is it OK to discuss such a specific electronic topic here?

Yes, of course. Try and keep each top-level discussion MicroPython-relevant, but absolutely it is common for the discussion to go into the details of the electronics.

**Comment by volkerjaenisch** [comment]
> This is news to me ... which bits of hardware stay powered-up?
I do not know. But the current my board is drawing in deep sleep (4mA)  is higher than expected.
Therefore I came up with the idea to talk about this topic.

For instance I2C can be initiated by Micropython but there is no corresponding deinit() function as there is for other components like e.g. WLAN or UART. This asymmetry I do not understand.

Cheers,
Volker

**Reply by jimmo** [reply]
> For instance I2C can be initiated by Micropython but there is no corresponding deinit() function as there is for other components like e.g. WLAN or UART. This asymmetry I do not understand.

I forget exactly the conclusion from last time we talked about it, but I think this is something we need to add. It's just a matter of someone putting the time into implementing. For example it's also a good way to let the pins go back to their default state.

**Comment by davefes** [comment]
With a ESP32-WROOM DevkitC board, 3v3 regulator removed and the supply cut to the CP2102 I am under 1mA in deepsleep.
Provide a link to your actual hardware.

**Reply by volkerjaenisch** [reply]
https://www.dfrobot.com/product-1634.html
The voltage regulator used on this board is a RT9080-33GJ5, which is quite efficient.

**Reply by davefes** [reply]
Looking at the picture and reading "programming via Micro-USB" suggests to me that the lil chip in the bottom right hand corner of the ESP8266 module is a UART to USB bridge.  The CP2102 on my boards draw about 3-4mA.  

I mostly use bare ESP32 modules and a break-out board to get around the hassle of the disconnecting and re-connecting power to the CP2102 every time something goes wrong.

**Reply by volkerjaenisch** [reply]
Stupid question: Is there a way to use "bare ESP32 modules" without SMD soldering (which I am not capable of [Bad eyes])?

**Reply by davefes** [reply]
If you click on the links and use Aliexpress you might need to reset your home country.

I buy these:
https://www.aliexpress.com/item/32961594602.html?spm=a2g0o.order_list.order_list_main.196.23771802pVEBGJ
these:
https://www.aliexpress.com/item/32810607598.html?spm=a2g0o.order_list.order_list_main.45.23771802pVEBGJ
and then mount the ESP32 bare modules on these:
https://www.aliexpress.com/item/1005004916288484.html?spm=a2g0o.order_list.order_list_main.39.23771802pVEBGJ
I was just able to solder wires on the bare modules but making changes when the module was mounted on a piece of PCB material and inside a box was difficult.

You have to index the module accurately on the break-out board but soldering it is a lot easier then soldering individual wires on the bare module.

Be aware:  you have to make up a programming harness with a boot and reset button as programming the bare board is not as straight-forward as the dev boards.

**Reply by volkerjaenisch** [reply]
Thank you so much for the information.  But Aliexpress complains that these products are not available (probably because I am in Germany).

**Reply by davefes** [reply]
On Aliexpress try searching for:
`ESP-32S Breakout Board Converter Module`
If you use these I would suggest a dab of heatsink paste on the little square in the middle before mounting or better still reflow the module on to the carrier.

**Reply by volkerjaenisch** [reply]
Yeah! Got it! Ok, this I can solder.
Sorry for being such a NOOB, and thank you for the support.

**Reply by davefes** [reply]
I have been building hardware for 60 years and just discovered these particular break-out boards, so I would be surprised that everybody knew about them :)

**Reply by volkerjaenisch** [reply]
You are welcome.

**Reply by glenn20** [reply]
Or, there is this monstrosity I built a few years ago when prototyping my own low power esp32 devices... :wink:. It's actually not hard to solder onto the bare modules.

![IMG_20200911_182100~2](https://github.com/micropython/micropython/assets/6965319/86470776-591a-4c00-a6b7-b7b2603db5cf)

I even soldered smd LDOs and caps directly onto the ribbon cable. (BTW. - not a recommended solution, but very compact).

**Comment by volkerjaenisch** [comment]
> > each bit of the ESP32 hardware has to be shut down manually by the programmer before entering deepsleep. This fact is not documented.
> > And it is not really clear how to invoke a hardware component and how to set it to sleep.
> 
> but other sources of information are
> 
>     * the ESP-IDF documentation
> 
>     * data sheets
> 
>     * source code.
But what costs the sentence "If you set your board to deep-sleep, please make sure that all components you have invoked are shut down first." in the documentation.
> 
> I really have no idea why so many people seem to expect that uPy docs should cover every possible hardware configuration under the sun. Programming bare metal (or freeRTOS) **does** require at least minimal knowledge about hardware, protocols, etc.

1) I do not understand why you are reacting so aggressively to my friendly proposal. 
2) Your response does not contain any usable information beside the rant : "Use the force read the source".

And yes, you are right. I have only a physics diploma but I have no deep electronics knowledge. And therefore I come from a different context. And we all know how important context is. I am a Programmer for 40+ year starting with machine language on a SHARP PC1500 and I am now developing applications. And my context knowledge is that when I enter a deep-sleep on e.g. my linux-system the OS will shut down the hardware. And I wrongly assumed that MicroPython would do the same thing on a ESP.
The above mentioned additional hint in the deep-sleep documentation would have given me the correct context and spared me and others a lot of red herring chasing.

 
3) I gave an concrete example that e.g. for I2C there is a deinit() only for certain boards. And your answer is simply:
> This is not true. 

For that bold and absolute statement you gave zero reason.

Can you please be in the future a bit more constructive.

Cheers,
Volker

**Comment by volkerjaenisch** [comment]
Dear @glenn20 !
> I just ran a lightsleep and deepsleep test on my FeatherS3 with the PPK2:
> 
>     * Normal idle operation (wifi off): **46mA**
> 
>     * Light sleep:  **1.3mA**
> 
>     * Deep sleep: **0.043mA**

That is quite impressive. Many thanks for your demonstration.

Now I only need a distributor that sells/ships a FeatherS3 in/to Germany. Any hints anybody?

Pimoroni where I bought Adrafruit products in the past is "out of stock" and also their product pictures for the FeatherS3 do not show the second Stemma QT-Port. Are there different variants/versions of the FeatherS3?

Cheers,
Volker

**Reply by peterhinch** [reply]
There are two variants, one with an onboard antenna, the other with a Ufl connector - see [this page](https://esp32s3.com/index.html#esp32s3_info). Both have two Stemma ports.

**Reply by volkerjaenisch** [reply]
So the product images are indeed wrong on the pimoroni website. Thank you for clarification.

**Comment by volkerjaenisch** [comment]
Dear @glenn20 
> I know that you are using an esp8266 device, but I think the best opportunities for achieving very low power operations are probably with the ESP32 devices (partly due to the availability of modern well-optimised boards like the feathers3 mentioned by Peter) and much more comprehensiive documentation (though often ambiguous).

The esp8266 was a bad buy (never trust amazon - I ordere a DFrobot ESP32 but got a 8266), but since it also utilized the  RT9080-33GJ5 I gave it a try.

But as @davefes found out the voltage regulator is not the problem on this board but probably the CP2102.

Just ordered a FeatherS3 for eye watering 34EUR via Amazon.uk.  Will be a fortnight to get it delivered. 
I will come back to share my experiences.

Cheers,
Volker

**Reply by glenn20** [reply]
If they are clever, the cp2102 may not be powered when running off lipo.

**Reply by glenn20** [reply]
> Just ordered a FeatherS3 for eye watering 34EUR via Amazon.uk. Will be a fortnight to get it delivered.

I have the good fortune that these devices are built only a few suburbs from where I live ;).

**Comment by UnexpectedMaker** [comment]
@volkerjaenisch I ship directly to DE from Australia -  Either from my website, or from Lectronz (that handle the VAT/Fees for you).
https://lectronz.com/products/feathers3

Issue shipping from Australia is the insane postage prices since Covid that unfortunately will never go back down :(

BTW, Pimoroni have the correct pictures for the FeatherS3 on their site - maybe you were looking at the FeatherS2 instead?
https://shop.pimoroni.com/products/feathers3-esp32-s3?variant=39762596298835

Cheers :)

**Reply by volkerjaenisch** [reply]
@UnexpectedMaker  Thank you so much for your support, I am overwhelmed!
Will look how the Amazon-UK delivery works. Then lectronz I will try as the next option.

The EU CE certification is the new fourth rider of the apocalipse. It brings fear, desease and decay onto us makers.

I developed a device to fotograph and classify rain filter probes for the German Weather Service. All of its components (RPi, industrial camera, power supply, servos, calibrated color card) were CE certified. My added work was some 3D-printed gears, mechanics and housing (and a lot of Software). But the whole thing had to be anew CE certified. Luckily I found a hole - instruments build for scientific research do not need a CE sign. But only for 3 years duration.

(Edited by @jimmo to remove a paragraph that was not appropriate)

Cheers,
Volker

**Comment by UnexpectedMaker** [comment]
The FeatherS3 should be able to get down to < 30uA deep sleep - but it depends on what you have connected to it, and how you are waking it. I've had it as low as 20uA @ 3.3V, and 26uA at 4.2V, both with nothing connected to it.

But honestly, anything under 50uA it perfect for deep sleep current - The biggest cost is always going to be when the board is awake and the radios are either calibrating, or you are transmitting data.  A super low deep sleep current is only helpful if you plan on having the board sleep for days at a time without waking, and then you have the issue of RTC drift with the internal ESP32 RTC :(

**Reply by volkerjaenisch** [reply]
Dear @UnexpectedMaker !

The FeatherS3 arrived today. But I have stability issues (serial drop outs) or do not understand the Feather correctly.
Shall I elaborate here or open a new thread?

Cheers,
Volker

**Reply by volkerjaenisch** [reply]
The stability issues were due to my USB setup. Obviously the ttyACM does not like my USB-hub. Without hub the featherS3 runs quite normal.

But one question concerning deep sleep. When my code enters deepsleep my IDE (Thonny) looses connection to the featherS3?

When I issue on the esp32 command line a deepsleep(x) the featherS3 goes to sleep and after wakeup the terminal reconnects.
When the deepsleep is issued from my code the terminal does not reconnect after wakeup?

Can you please shed some light?


Cheers,
Volker

**Reply by UnexpectedMaker** [reply]
There are USB hub issues with the ESP32-S3 when using native USB - It's not specific to my boards, and largely depends on the Hub. I'm not sure what Espressif are going to do about it, or if they can at this stage.

With regards to losing connection with DeepSleep, that is the correct behaviour - native USB can't work when the device is sleeping, as native USB runs in the main cores.

That is why Thonny loses connection as the USB is shut down when sleeping.

`When I issue on the esp32 command line a deepsleep(x) the featherS3 goes to sleep and after wakeup the terminal reconnects.
When the deepsleep is issued from my code the terminal does not reconnect after wakeup?`

I don't understand this, sorry. You mean when you issue a command from REPL?  as opposed to just booting your code?

Working with native USB is quite different to using a SerialUART - what most people are used to. Remember, when you are running off battery, it doesn't matter what USB is doing :)

**Reply by volkerjaenisch** [reply]
Thank you for the fast reply and the light you shed. 

> I don't understand this, sorry. You mean when you issue a command from REPL? as opposed to just booting your code?

Sorry for my bad English. Yes, this I meant.

And of course you are right that the USB is unimportant during the production. But I like to see what the code is doing while developing and the USB is a nice and easy possibility.

My logger class can also log to the UART so I do not need the USB.

Tomorrow I will make some measurements on power consumption, so please stay tuned.

Cheers,
Volker

**Reply by volkerjaenisch** [reply]
I am still running into problems with the FeatherS3. 
Board works fine on USB for several hours.
But on Battery: 2AH LiPO it is not capable to connect to the net. Looks like a brownout.
My vacation has run out and so the time for this project.
I will infrequently update.

@all: Many thanks!

Cheers,
Volker

**Reply by UnexpectedMaker** [reply]
Looks like a brown out?  From the behaviour, or because you actually see a brownout confirmation in the boot logs via TX/TX ?
 
It's very unlikely a brownout if it doesn't brown out from 5V power source, as it's the same path in, so it if is actually a power problem, it's likely noise being caused by something externally connected to the FS3 and powered from the FS3.

If you do find time to update us, I'd love to see a simple diagram of how everything is connected to the FS3.

**Reply by davefes** [reply]
How are you connecting the Li-ion to the board?  Via a regulator or diode to 3V3?  You could try a 1000uF low-ESR cap on the 3V3, but there are some conditions that you need to be aware of if going to light/deepsleep.

**Reply by UnexpectedMaker** [reply]
There's a JST connector on the board for the battery and a VBAT pin on the header, and they are the ONLY way the battery should be connected.

**Reply by davefes** [reply]
Noted.

**Reply by volkerjaenisch** [reply]
Dear @UnexpectedMaker !

Yes it is my guess only. The RX/TX schows no brownout but a lot of gibberich I cannot interpret.

The LiPo is connected via the JST header.

My code is quite to complex to use it for testing. 
I will strip it down to the parts that are not processing with the battery plugged in.

Cheers,
Volker

**Reply by volkerjaenisch** [reply]
Dear @UnexpectedMaker !

How are the hardware UARTs located in Pin-Space on the Feather S3?

I am using UART 1 with RX=17 and Tx=18 for my custom logging. Can this be the problem?

Are there guidelines how to debug such a situation? If not please give me hints.

And to state it again: My code runs fine and stable for hours powered by USB. But it fails instantly while battery powered. 

The Rx/Tx log from the board is not really readable. Here I need some pointer to improve.

QTTI�PQ�IDLD�DDI��1323
�EZ��
     Ma 21 201
214& (BRCQFAUTCRST),bcc4 (SPIABASTCFLASHCBAAT)
SPIQP
     0ee
        -cdeDIA, clc!k $i1
                           laad01fae0,leff34
d!!$
    01!100,de&0c4
                     ,cad403cc300,lef02a
                                          ef42   403c&4
�QTTI�PQ�Y&DD�DDI��123
�EZ��
     Ma 1 2
                214& (BRCSFAUTARST),bcc (SPIAFASTCFLAQHCBCCT)
SPIQP
     0ee
        -adeDIA, clcck di61
lcad3fce30$lef0f14
lcad0403c100,lef0c44
                     lcad0403cc300,le&02c3
%&42 403c&4
 QTTI��Q�IDLD�DDK��103
                         �EJ��
                              Ea 21 21
214& (BRCSFCUTCRST),b#a4
                         0 (QPICFASTCFLASHCBAAT)
SPIQPe%
        mcdeDIA, clcck di61
,cad03fce30,lef0f34
dca$
    03c100,lefc
                    ,!!d401cc0$leb02c
%&42     401af4
                 �Q*TI�PS�Y&DD�DDK��1032
�EZ��
     Ma 21 201
214
   0f (BRCSFCUTCRST),bc#40 (SPICBASTCFLASHCBACT)
SPISP%%
mcdeDIC, clcck di61
lcad03fce0,leb0f34
lca$
    0403a30,lefc
dc!d001cc0$$%b2!1
ef42 001f4
            QTTK�PQ�IDDD�D�I&D�DDD�DC��EZ��
                                           Ma 23 2021
2140f (BRCSFCUTCRST),bcc40 (SPIAFASTCFLASHCBCCT)
SPISP0ee
-cdeDIA, cdcck di1
                   lcad03fce30,lef0&34
lcad003c10,lef0!44
d#!$
    040!a10,le"02a3
%f42 401cf4
              �QTTI�PQ�IDLD�DDI��1023
�EZ��
     Ma 21 2021
2340f (BRCQFCUTCRST),bcc40 (SPIAFASTCFLASHCBACT)
SPISPe%
mcdeDIA, cdcck di61
laad01fce30,lef0f34
l!!d
    043c10,def0c4
                     ,cad040cc300,lef02c3
                                          e&42 0403cf4
QTTI�PQ�IDLD�DDI��10323
�EZ��
     Ma 23 201
2340f (BRCSFCUTCRST),bcc0 (SPICFASTCFDASHCBCCT)
SPISP0ee
        m!deDIC, clcck $i61
                           d!!d03fc%30$$%f
                                          f34
lca$4c100,lef0!44
lcad003cc100,lef02c3
                     ef42 0cf4
QTTI�PS�YDDD�DDI��1023
�EZ��
     Ma2 21 21
240f (BRCQFAUTCRST),bcc (SPIAFASTCFLASHCBCAT)
SPIQP
     e%
        -adeDIA, clcck di61
,aad03&ce30,lef0f34
lcad003!300,lefc44
,aad0403cc300,lef02c
%&42 0403cf4
QTTI�PQ�YDLD�DDI��123
�EZ��
     Ma2 21 2




Cheers,
Volker

**Reply by UnexpectedMaker** [reply]
That looks like a baud rate issue, there should be no issues using any IO as HW UART pins on the ESP32 - you could just use UART0 that are TX/RX on the headers. They should already be dumping boot info as the ESP32 starts up.

**Reply by volkerjaenisch** [reply]
The connection diagram:

Sensor -> Feather: 
GND -> "-"
VCC -> 3.3V LDO2
SDA -> SDA
SCL -> SCL

Serial Code:
UArt1: RX=17, TX=18

Cheers,
Volker

**Reply by UnexpectedMaker** [reply]
How do you know it's not working on battery? Are you using the RGB LED or something to indicate the wifi status? Is it just the wifi connection that is failing?

**Reply by UnexpectedMaker** [reply]
Are you on discord? Maybe at this stage, it might be easier to troubleshoot this via my Discord server?
https://discord.com/invite/DZPA5rgyGt

**Comment by robert-hh** [comment]
> Now I only need a distributor that sells/ships a FeatherS3 in/to Germany. Any hints anybody?

exp-tech.de have them on their web site. You have to register as company first, and then they ship it. They stopped shipping to non-commercial customers. Maybe they got tired of people shipping back their purchase.
Eckstein-shop.de as well, but with 2 weeks delivery time, indicating that they buy it for you from US on request. In that case it uses to fail.

Note: I had bad experience with shops from the US promising to handle VAT/customs for you. That failed, and I had to pay VAT/Customs twice + handling fee. You can reclaim that, but it is a formal process which takes weeks. That whole customs stuff only worked for me with Aliexpress, Mouser & Digikey.

**Reply by UnexpectedMaker** [reply]
Neither of those companies are buying wholesale from me - They are just buying via Adafruit when they stock other Adafruit boards, so their price is much higher than it should be. That's ok, if the lower shipping rates even it out :)

I can't vouch for VAT pre paid from US vendors - but I use Lectronz for EU orders for those that want the VAT taken care of as part of the purchase, and Lectronz are an EU based company, and they do indeed manage the VAT etc.

I've not had any customers that bought via the Lectronz store have any VAT/Duty/Fee issues on delivery.

Seon

**Reply by UnexpectedMaker** [reply]
The reason they only do B2B (and don't sell to consumers) is it's illegal to sell them in the EU to consumers as they are not CE certified.  B2B is ok though, so that is why they require business sign-up/registration.

It's why I don't have any EU distributors - Silly EU :(

**Reply by robert-hh** [reply]
Yes, Mouser, Exp-Tech and sell them as Adafruit products. And Exp-Tech does B2B for all of their products, even those manufactured in DE. They just stopped selling to end customers. Maybe end customers returned too many products.

> Lectronz are an EU based company, and they do indeed manage the VAT etc

If that works, that's fine. I just looked it up at Lektronz: €44.26 (incl. €17.21 shipping and €7.06 VAT/taxes). This whole customs stuff is so medieval! I stopped purchasing from UK companies since Brexit, just because of that additional trouble.

**Reply by UnexpectedMaker** [reply]
Yeah, the shipping costs from Australia are really bad :(
A wonderful country to live in, but a terrible country to export goods from !

**Comment by volkerjaenisch** [comment]
> Looks like a brown out? From the behaviour, or because you actually see a brownout confirmation in the boot logs via TX/TX ?
> 
> It's very unlikely a brownout if it doesn't brown out from 5V power source, as it's the same path in, so it if is actually a power problem, it's likely noise being caused by something externally connected to the FS3 and powered from the FS3.
> 
> If you do find time to update us, I'd love to see a simple diagram of how everything is connected to the FS3.



> That looks like a baud rate issue, there should be no issues using any IO as HW UART pins on the ESP32 - you could just use UART0 that are TX/RX on the headers. They should already be dumping boot info as the ESP32 starts up.

The log output was from the TX/RX on the header.

I will look into this - but its 4:00 local time - time to sleep.

Cheers,
Volker

**Comment by xyzzy42** [comment]
I've also got a UART issue related to power management on the ESP32-S3 that looks very similar to this.  I've tracked it down to the UART changing baud rates when the frequency is set to 40 MHz.

If pm config is set to max freq 160 MHz, min freq 80 MHz, then the UART works just fine.  Set it to max 160 and min 40, then it stops working.  It's not entirely broken, some data comes through, especially when code is running, like at boot up.  But at an idle repl prompt when BLE or wifi is not enabled, it doesn't work at all.  Except it does work, the baud rate is just half what it should be!

I haven't checked to see if I can replicate this in a pure ESP-IDF based app or if it only happens in micropython.
